### DISCUSSION TITLE
YAI with ILI9488.  I could use a little direction.

### DISCUSSION THREAD

**Discussion by kwpodolsky** [Bodmer/TFT_eSPI#2769] [Q&A - General]
I'm sure you are getting tired of responding to the ILI9488 SPI issues, but
I could really use some help.

I designed a board that uses the ESP32S3 Devkit C module and connects to a header
setup to directly interface with either the 2.8 or 3.5 TFT displays.
The 2.8 uses the ILI9341 and the 3.5 uses the ILI9488, but the pin definitions
are EXACTLY the same.

I created two USER header files for each of the boards.
The pin definitions are EXACTLY the same (I've attached them for viewing).

Problem is... when I setup the ILI9341 and put the display on,
EVERYTHING works fine, including calibration.

When I switch over to use the ILI9488, the display works fine but
NOT the touch.
I've seen your other posts about NOT connecting SDO(MISO) to the display.
I tried that... when I run the Calibration app, I do get a display but, it just runs thru the four touch points
and gives me some random data back for the calibration points.

Can't seem to get this to work.  If you could give me a push in the right direction
I would greatly appreciate it.

I've included a couple of screen shots of my board schematic, so you can see
the pins being used match the USER_SETUP files.

Again, thanks in advance.

![USER_SETUP__ILI9341](https://github.com/Bodmer/TFT_eSPI/assets/92733023/41c96d62-6798-4c8c-8001-da02ffd759c5)
![USER_SETUP__ILI9488](https://github.com/Bodmer/TFT_eSPI/assets/92733023/fcc955d4-3a92-4dfd-95aa-a74125c6795e)
![ESP32S3_PinUsage](https://github.com/Bodmer/TFT_eSPI/assets/92733023/e188543d-07d0-40eb-a3fe-164778d2f31a)
![SPI_PinHeader](https://github.com/Bodmer/TFT_eSPI/assets/92733023/deac49a7-74c8-474e-9846-30d7041fa4a4)

**Comment by TangerineDreamer** [comment]
Hi,I suggest you do another calibration because those 2 displays are not the same dimension and the touch is bad for one,
you'd better erase the SPIFFS where the last calibration was registered before you do a new calibration,why not begin
with the ILI9488 first.
Could be the TOUCH_CS_PIN maybe not work with ILI9488,have you tried it before you designed your board?

**Comment by kwpodolsky** [comment]
Hi Bodmer/TFT_eSPI

Thanks for the reply.
Yes, I realize the dimensions are different and yes, I erased the SPIFF
every time before I load new calibration code for either display.
So, I don't think it's that.

As for the TOUCH_CS pin...
It works FINE on the ILI9341, so not sure why it wouldn't work for the 9488
(logically it should, the processor doesn't care. That pin is as good as
any other).

No, I didn't test it, with the 9488, while designing the board.
Unfortunately, I didn't have one (it was on order), but since all the
pinouts on the 9488 were the same as on the 9341 I...
ASS-UMED...
that the 9488 would function with the same pins as the 9341 (and why
wouldn't it?).

It makes no logical sense that it wouldn't, but....

If you have a pinout design and a snippet of code you could share with me
of how you did it... maybe I can figure it out.

Thanks again.
Kerry

**Comment by TangerineDreamer** [comment]
i'm not bodmer ! look at my pseudo :)
I often go on the forum when there are tft troubles, if i can help a some,
maybe you can try:
TOUCH_CS 12// (i use it for ILI9488)
TOUCH_CS  27//(I use it for ILI9341)
maybe one those works for the twice

**Comment by kwpodolsky** [comment]
Sorry about that.
I didn't actually go to the web site, I just responded to the email,  so I
ASSUMED (again) that you were Bodmer.

I could try the changes you mentioned, but WHY?
Is there something more special about GP12/27 on the ESP32S3, then there is
the GP14 pin, for the TOUCH_CS?
Something you know that the datasheet doesn't say?

As I said before, the board works (as designed) with the 9341.  Question
is, why doesn't the 9488 work the same?
I get that Bodmer has stated, don't connect SDO (MISO) line... but why
doesn't MY touch work?
And why if I don't connect SDO, does the code just run thru the test points?

Thanks for trying to help.
But I need something concrete, a reason.
I've been programming professionally since 1977, so I don't like not
knowing WHY something doesn't work... when it should.

;-)

**Comment by TangerineDreamer** [comment]
I don't have any esp32 s3,only regulars esp32 devkit1,so the setups i collected were the results of my experiencing
with my esp32's,but i will try to verify tomorrow if any TOUCH_CS PINS could work.
Have you ever thought your ILI9488 was maybe a little broken or have you tested it on the fly with another esp32
or else mcu?
I own a ST7796 wich have a definitively not "working at all" touching function and i realized it after wasting days to test many
wirings and different sketches.

**Comment by kwpodolsky** [comment]
I have TWO 9488s and neither work.
So I'm fairly sure it's something fundamentally wrong, that either I'm
doing or NOT doing. 😁

from my Pixel 5a 5G with Android 11

**Comment by TangerineDreamer** [comment]
It's a pretty weird situation,
an old chinese man said: "when you reach the end of the rope,make a node and hang on to it..." 

Did you do this: erase the spiffs (the FS erase  flash) and set the calibration to "true" at the beginning of the sketch.

**Comment by TangerineDreamer** [comment]
I've made today a lot of tests of TOUCH_CS  with my esp32devkit1 and ili9488 & ili9341; 
          (CS 12 & DC 14 & RST EN); (MOSI 23,MISO (nothing),TCLK 18 , Touch_DO 19)

first i discovered TOUCH_CS works everywhere exept PIN 34 & 35 and curiously can works at the same pin of CS = 12 !!
Then i put on ili9341 with the ili9488 setup and i can see the display but with wrong colors and dimensions,but the
touch is the same as ili9488,so the calibration is not an obligation,st7796 is totally compatible with ili9488 setup and touch.
But if you put on ili9488 on the ili9341 setup,you see nothing.
I think it's interesting for designing if you can wire CS and TOUCH_CS  together.

**Comment by Bodmer** [comment]
1. Make sure pin 9 of SPITFT connector is not connected with ILI9488 because that TFT pin does not tristate.

2. Run the Test_Touch_Controller sketch and see what is output when the screen is touched and not touched. Post part of output here.

**Comment by kwpodolsky** [comment]
As per my initial post...

1. SDO(MISO) Pin 9 NOT connected to 9488 (with the 9488 driver).
2. ERASED the SPIFF
3. Ran the Test_Touch sketch.
4. Results... Screen initializes, I see the text, I see the test points on
the screen...
the software does NOT wait for the screen to be TOUCHED.
It just runs thru all the test points and then posts up some bogus array of
data points.

BTW, this happens with TWO different 9488 displays.  So, it is something
systemic, not a display issue.

When I perform the EXACT same steps (excluding #2) with a 9341 display
attached (and the 9341 driver)... EVERYTHING works as expected.

WHAT is so drastically different about the 9488 from the 9341 controller?

**Comment by kwpodolsky** [comment]
Please read my initial post... I am NOT using a standard ESP32 DevKit board.
I am using an ESP32S3 DevKit C board.

Actually the calibration is necessary, since the 9341 is 320x240 and the
9488 is 480x320.

When I run the code (with the proper driver) on either display...
Everything displayed is FINE, it is the TOUCH portion that
does not work properly when using a 9488 display.

If you wired the two CS lines together you would have an issue... the 9488
display and the Touch controllers would be confused with what is being sent.

**Comment by TangerineDreamer** [comment]
I think that the sketch, bodmer was talking about, only use the serial monitor,no tft displaying,so you don't have to see the
four corner dots on the screen:

[TFT_eSPI](https://github.com/Bodmer/TFT_eSPI/tree/master)/[examples](https://github.com/Bodmer/TFT_eSPI/tree/master/examples)/[Test and diagnostics](https://github.com/Bodmer/TFT_eSPI/tree/master/examples/Test%20and%20diagnostics)/[Test_Touch_Controller](https://github.com/Bodmer/TFT_eSPI/tree/master/examples/Test%20and%20diagnostics/Test_Touch_Controller)

 ->then you only see numbers on the monitor as  x & y coordonates.

Well my tests only concern esp32devkit1 and i succeeded with CS & T_CS together with the" keypad" sketch
,one else can verify if he wants to.
Another thinking: when you bought those 9488,did you really choose the touch option?
I know aliexpress vendors propose that and i always be sure to what i choose.

**Comment by kwpodolsky** [comment]
Turns out, you might be right.

BOTH 9488 boards I have do NOT have a U2 chip, whereas the 9341 displays DO
have the chip.
I'm assuming that is the touch controller chip.
So, I've been spinning my wheels with this for a while, when the boards
don't even do touch.

Just ordered new boards from Amazon that showed the chip on the display as
well as says... TOUCH in the description.

Thanks for pointing this out to me.  I just thought all these displays had
touch built in... why else have the pins for it?

**Comment by TangerineDreamer** [comment]
Look at that:

https://fr.aliexpress.com/item/1005005385630017.html?spm=a2g0o.productlist.main.15.503c34e1EqwQ68&algo_pvid=20396ebc-879a-4cd7-b9f6-5675eee0d2bd&aem_p4p_detail=2023070101300013841331922982100005190734&algo_exp_id=20396ebc-879a-4cd7-b9f6-5675eee0d2bd-7&pdp_npi=3%40dis%21EUR%2114.15%2112.74%21%21%21%21%21%402100b78b16882002009444031d07b2%2112000032838895395%21sea%21FR%21764226657&curPageLogUid=EC0YkP4uAYiX&search_p4p_id=2023070101300013841331922982100005190734_8

you can select whether you want with touch or not, they will not make a special hardware for the no-touch,too much
expensive to build :)
did you try the sketch bodmer pointed to?
I did'n understood when you said "Turns out, you might be right".
Can you show the amazon link?

**Comment by kwpodolsky** [comment]
What I meant by... "Turns out, you might be right"... is that
The boards I received do NOT have touch capability.
The U2 chip (which I believe is the touch controller) is missing on my 9488
displays, but is present on the 9341 displays.

I ordered new boards from Amazon that have the U2 chip and the description
does say TOUCH capable.
So, I should have them in a couple of days and I will try my setup on the
new boards.

I will post back with my results.

Thanks for pointing out that not ALL boards are touch ready.

**Comment by TangerineDreamer** [comment]
Ok tell us when you will realize it with success :)

**Comment by Bodmer** [comment]
Note that screens have also been supplied that have the touch controller chip (XPT2046) fitted but *NO* touch sensitive film attached to the front of the display. So check both are present otherwise touch functions will not work.

It is important that two chip select signals are used in the case where the display driver chip and touch controller chip are on the, one for the display and one for the touch controller.

**Comment by TangerineDreamer** [comment]
> It is important that two chip select signals are used in the case where the display driver chip and touch controller chip are on the, one for the display and one for the touch controller.

Hi bodmer,how can you explain touch works while CS & T_CS are on the same pin,what are the risks?

**Comment by kwpodolsky** [comment]
Ok, I got two new 3.8" ILI9488 TOUCH screens and I've gotten them to work with my ESP32S3 Devkit C1 boards.
NOW, the only problem is, I lost usage of my SD card because.... it uses the same SPI lines (except the CS) as the display and...
As Bodmer has pointed out... the 9488 does NOT play nice with other devices using the same SPI lines.
So... two steps forward... one step backward.
I now need to figure out HOW to implement SPI3 on the ESP32S3 to support the SD card separately.

Anyway, thanks for the support from the group.

**Comment by TangerineDreamer** [comment]
Is was very fast to get new 9488!!
Look at this post,you'll have a solution to try :)

https://github.com/Bodmer/TFT_eSPI/discussions/2717

**Comment by kwpodolsky** [comment]
Thanks, I'll check it out.

from my Pixel 5a 5G with Android 11

**Comment by Bodmer** [comment]
Assumed resolved.
